game/final_obs_gen.py

```diff
@@ -0,0 +1,494 @@
+                    [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0],
+                ]
+            )
+
+    def setup_parameters(self):
+        self.height, self.width = self.ARRAY.shape
+        if self.obs_type in ("c", "C"):
+            self.y = self.REF_AXIS - self.height * self.GRID_SIZE
+        elif self.obs_type == "b":
+            self.y = self.LOW_BIRD_AXIS - self.height * self.GRID_SIZE
+        else:
+            self.y = self.HIGH_BIRD_AXIS - self.height * self.GRID_SIZE
+
+
+class ObstacleInteraction(Game):
+
+    def __init__(self):
+        super().__init__()
+
+        self.curr_cooldown = 0  # generates after however long (IN PIXELS)
+        self.cooldown_variance = np.linspace(10, 50, 5)  # variance to cooldown time
+        self.base_cooldown = 40  # must wait this long before new cacti generates
+        self.curr_obs = []  # obstacles generated for the game
+        self.possible_speeds = [10]
+        self.possible_objects = ["c", "C", "ccc", "CC", "cCc", "cc", "b", "B"]
+
+    def update(self):
+        for i, obs in enumerate(self.curr_obs):
+            if obs.x + obs.width > 0:
+                obs.x -= obs.speed
+            else:
+                self.curr_obs[i] = ""
+        if "" in self.curr_obs:
+            self.curr_obs.remove("")
+
+    def make_obstacle_choice(self):
+        # choose between cactus arrangement, low bird, or high bird
+        return np.random.choice(self.possible_objects), np.random.choice(
+            self.possible_speeds
+        )
+
+    def generate_obstacle(self):
+        if self.curr_cooldown == 0:  # time to generate an object
+            new_obstacle, new_speed = self.make_obstacle_choice()
+            for i, obs in enumerate(new_obstacle):
+                new_obs = Obstacle(obs, new_speed)
+                # Note: Need to figure out correct line below to get the cacti close together
+                new_obs.x += (new_obs.width + 1) * i * new_obs.GRID_SIZE
+                self.curr_obs.append(new_obs)
+            self.curr_cooldown = self.base_cooldown + random.choice(
+                self.cooldown_variance
+            )
+        else:
+            self.curr_cooldown -= 1
+
+
+class Dino(Game):
+
+    def __init__(self):
+        super().__init__()
+
+        # Jumping Constants
+        self.U_JUMP_HEIGHT = 30  # dino is 20x19 in size
+        self.JUMP_TIME = 0.6  # full time to rise up and fall back down, seconds
+
+        # Run Initiation
+        self.setup_sprites()
+        self.setup_parameters()
+
+        # Game Attributes
+        self.ARRAY = self.JUMP_ARRAY
+        self.curr_walk = self.UPDATE_WALK_FRAMES
+        self.x = 2 * self.GRID_SIZE
+        self.y = self.OG_DINO_Y
+        self.is_right_foot_down = True
+        self.is_jumping = True  # Detects if jumping signal is given
+        self.jump_start_time = 0
+        self.is_ducking = False  # Detects if ducking signal is given
+        self.jumping_coefficient = self.JUMP_HEIGHT / (

```

Comment:
The reason we made it a constant was just not to need to calculate it every loop so that we can just define in terms of the height and time, which are more understandable